### PR TITLE
add WebTransport multiaddr components

### PR DESCRIPTION
**Pull request by marten-seemann**
Based on https://github.com/libp2p/specs/pull/404.

### CODE REVIEW

File: transcoders.go [hunk 1]
```diff
@@ -373,3 +374,14 @@ func dnsStB(s string) ([]byte, error) {
...
+
+var TranscoderCertHash = NewTranscoderFromFunctions(certHashStB, certHashBtS, nil)
+
+func certHashStB(s string) ([]byte, error) {
+	_, data, err := multibase.Decode(s)
+	return data, err
+}
+
+func certHashBtS(b []byte) (string, error) {
+	return multibase.Encode(multibase.Base58BTC, b)
```
Comment (marten-seemann) [inline transcoders.go:386]: This choice is totally arbitrary. We could use any other multibase that's widely supported.

Is there any best practice for choosing one over the other?

Comment (aschmahmann) [reply transcoders.go:386]: cc @lidel, but the general guidelines I know of are:
- Do you want compact -> choose base64URL (`u`), although base58btc is probably fine too
- Do you want lowercase -> choose base32 (`b`)

File: transcoders.go [hunk 2]
```diff
@@ -373,3 +374,14 @@ func dnsStB(s string) ([]byte, error) {
 func dnsBtS(b []byte) (string, error) {
 	return string(b), nil
 }
+
+var TranscoderCertHash = NewTranscoderFromFunctions(certHashStB, certHashBtS, nil)
+
+func certHashStB(s string) ([]byte, error) {
+	_, data, err := multibase.Decode(s)
```
Comment (aschmahmann) [inline transcoders.go:381]: Are there any other requirements here? e.g. is it a multihash, is it just 32 bytes, etc.

File: multiaddr_test.go [hunk 3]
```diff
@@ -157,6 +158,8 @@ func TestConstructSucceeds(t *testing.T) {
 		"/ip4/127.0.0.1/tcp/1234",
 		"/ip4/127.0.0.1/tcp/1234/",
 		"/ip4/127.0.0.1/udp/1234/quic",
+		"/ip4/127.0.0.1/udp/1234/quic/webtransport",
+		"/ip4/127.0.0.1/udp/1234/quic/webtransport/certhash/zt1Zv2yaZ",
```
Comment (aschmahmann) [inline multiaddr_test.go:162]: IIUC there can be multiple certhashes, if so we should probably test for multiple certhashes as well.

I'm not sure if the place is here or elsewhere, but wherever we're defining the semantics of `webtransport` and `certhash` we should add some tests around escaping (e.g. making sure you can do `.../webtransport/certhash/hash1/certhash/hash2/p2p/peerID` but not `.../webtransport/certhash/hash1/p2p/peerID/certhash/hash2`

Comment (marten-seemann) [reply multiaddr_test.go:162]: > IIUC there can be multiple certhashes, if so we should probably test for multiple certhashes as well.

Good point. Will add.

> I'm not sure if the place is here or elsewhere, but wherever we're defining the semantics of webtransport and certhash we should add some tests around escaping

It probably needs to live in the WebTransport implementation. Having one certhash, then something else, then another certhash can be a valid multiaddr (e.g. a circuit).

File: transcoders.go (hunk 2 above)
Comment (marten-seemann) [reply transcoders.go:381]: I added the validation on encoding.

File: transcoders.go (hunk 1 above)
Comment (marten-seemann) [reply transcoders.go]: Switched to `Base64url`.

**Review by mxinden — COMMENTED** [review]
:rocket:

File: protocols.go [hunk 4]
```diff
@@ -15,6 +15,8 @@ const (
 	P_IP6ZONE           = 0x002A
 	P_IPCIDR            = 0x002B
 	P_QUIC              = 0x01CC
+	P_WEBTRANSPORT      = 0x01D1
+	P_CERTHASH          = 0x01D2
```
[the remark this answers is not in the document]
Comment (marten-seemann) [reply protocols.go:19]: https://github.com/multiformats/multiaddr/pull/130

**Pull request merged**